### PR TITLE
make all classes haxe3 compatible

### PR DESCRIPTION
**Pull request by grepsuzette**
No idea if the team would find it worthwhile to have these new crypto classes also available for Haxe 3 (see https://github.com/HaxeFoundation/crypto/issues/6), in any case I wanted to post this simple PR.

### CODE REVIEW

**Comment by Simn** [comment]
Is there any point to this given that the Haxe 3 standard library has these classes anyway?

Well, I suppose it's nice to be able to do `-lib crypto` and have it just working with both Haxe 3 and 4. So I'm fine with making these changes.

**Comment by grepsuzette** [comment]
> Is there any point to this given that the Haxe 3 standard library has these classes anyway?

I am under the impression Haxe 3.4.7 haxe.crypto only features:

* Adler32
* Base64
* BaseCode
* Crc32
* Hmac
* Md5
* Sha1
* Sha224
* Sha256

Thus missing:  Aes, BCrypt, BlowFish, Des, Pbkdf2, Ripemd160, Sha384, Sha512, TripleDes, TwoFish.
I'll let you check if that's actually the case cause I'm not sure how to be sure, but for example Aes has very nice possible applications and Sha512 begins to become used in some area (e.g. signing on some cryptocurrency exchange APIs). Arguably all of them will have nice use cases.

**Comment by Gama11** [comment]
`#if haxe3 @:extern #else extern #end` is quite silly / ugly. At that point you should just use `@:extern` in both cases, it still works in Haxe 4.

**Comment by Gama11** [comment]
Besides, it doesn't actually do anything, `haxe3` is still defined in Haxe 4 since it just means >= 3. ;)

File: src/haxe/crypto/Sha224.hx [hunk 1]
```diff
@@ -27,9 +27,9 @@ import haxe.ds.Vector;
 */
 class Sha224 {
 
-    public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
         var sh = new Sha224();
-		var data = haxe.io.Bytes.ofString(s, encoding);
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
Comment (Gama11) [inline src/haxe/crypto/Sha224.hx]: I still don't think this works, given that `haxe3` is defined for both Haxe 3 and 4? `#if haxe4` should be better?

Comment (grepsuzette) [reply src/haxe/crypto/Sha224.hx]: Yes indeed. Thanks

File: src/haxe/crypto/Sha224.hx [hunk 2]
```diff
@@ -27,9 +27,9 @@ import haxe.ds.Vector;
 */
 class Sha224 {
 
-    public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha224.hx]:
```suggestion
    public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

File: src/haxe/crypto/Sha224.hx (hunk 1 above)
Comment (grepsuzette) [inline src/haxe/crypto/Sha224.hx]:
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

File: src/haxe/crypto/Sha256.hx [hunk 3]
```diff
@@ -28,9 +28,9 @@ import haxe.ds.Vector;
 */
 class Sha256 {
 
-	public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
+	public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha256.hx]:
```suggestion
	public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

File: src/haxe/crypto/Sha256.hx [hunk 4]
```diff
@@ -28,9 +28,9 @@ import haxe.ds.Vector;
...
 		var sh = new Sha256();
-		var data = haxe.io.Bytes.ofString(s, encoding);
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha256.hx]:
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

File: src/haxe/crypto/Sha384.hx [hunk 5]
```diff
@@ -3,7 +3,7 @@ package haxe.crypto;
 import haxe.ds.Vector;
 import haxe.Int64;
 import haxe.io.Bytes;
-import haxe.io.Encoding;
+#if !haxe3 import haxe.io.Encoding; #end
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha384.hx]:
```suggestion
#if haxe4 import haxe.io.Encoding; #end
```

File: src/haxe/crypto/Sha384.hx [hunk 6]
```diff
@@ -12,8 +12,8 @@ class Sha384
     public function new() {
 	}
 
-    public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
-		var data = haxe.io.Bytes.ofString(s, encoding);
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha384.hx]:
```suggestion
    public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

File: src/haxe/crypto/Sha384.hx [hunk 7]
```diff
@@ -12,8 +12,8 @@ class Sha384
...
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha384.hx]:
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

File: src/haxe/crypto/Sha512.hx [hunk 8]
```diff
@@ -3,7 +3,7 @@ package haxe.crypto;
 import haxe.ds.Vector;
 import haxe.Int64;
 import haxe.io.Bytes;
-import haxe.io.Encoding;
+#if !haxe3 import haxe.io.Encoding; #end
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha512.hx]:
```suggestion
#if haxe4 import haxe.io.Encoding; #end
```

File: src/haxe/crypto/Sha512.hx [hunk 9]
```diff
@@ -12,8 +12,8 @@ class Sha512
     public function new() {
 	}
 
-	public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
-		var data = haxe.io.Bytes.ofString(s, encoding);
+	public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha512.hx]:
```suggestion
	public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

File: src/haxe/crypto/Sha512.hx [hunk 10]
```diff
@@ -12,8 +12,8 @@ class Sha512
...
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
Comment (grepsuzette) [inline src/haxe/crypto/Sha512.hx]:
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

Comment (grepsuzette) [reply src/haxe/crypto/Sha512.hx]: (Sorry seems like I should have made a batch, first time I use this resolve feature)

**Comment by flashultra** [comment]
Thank you for your pull request.  I just did some test using haxe 3.4..4 and everything works .
 Here is example for Sha512 :
  https://try.haxe.org/#B7266   and  http://try-haxe.mrcdk.com/#0b8f9
I think it's safe to merge the changes.

**Comment by grepsuzette** [comment]
Maybe we should integrate automated unit tests for those, if not now at least soon afterwards.

**Pull request merged**